Default to BOLT 4 tlv payload format onions

ViktorTigerstrom opened the pull request:
Closes #1393.

Default to creating BOLT 4 tlv payload format onions for nodes for which we haven't received any features through node announcements, and where no other features are known such as invoice features nor features in the init msg for nodes we
have channel(s) to.

If there is a more convenient way in our testing utils to create a network of nodes, where the nodes to not announce their features through `NodeAnnouncement`s, instead of the test config function I created I'll of course change to that :)!

Happy to address any feedback!

ViktorTigerstrom commented on lightning/src/routing/router.rs:
```diff
@@ -1132,7 +1134,7 @@ where L::Target: Logger {
 					&node_info.features
 				} else {
-					&empty_node_features
+					&variable_length_onion_optional_features
```
I just want to clarify that I couldn't manage to create a test scenario where changing this specific line created a different result in terms of which payload format we create. Though I think it makes sense to change this line.

vincenzopalazzo replied:
```suggestion
					&onion_optional_features
```

ViktorTigerstrom commented:
Cleaned up some documentation.

TheBlueMatt reviewed (commented):
I don't believe this modifies the default for last-hop hints which do not appear in the network graph at all. This is the most common use for this. See the `NodeFeatures::empty()` call on L1317 in router.rs

ViktorTigerstrom commented:
> I don't believe this modifies the default for last-hop hints which do not appear in the network graph at all. This is the most common use for this. See the `NodeFeatures::empty()` call on L1317 in router.rs

Ah thanks, I'll update with converge of that case as soon as I can!

vincenzopalazzo commented on lightning/src/routing/router.rs:
```diff
@@ -1101,7 +1101,9 @@ where L::Target: Logger {
 	}
 
-	let empty_node_features = NodeFeatures::empty();
+	let mut variable_length_onion_optional_features = NodeFeatures::empty();
+	variable_length_onion_optional_features.set_variable_length_onion_optional();
```
```suggestion
	let mut onion_optional_features = NodeFeatures::empty();
	onion_optional_features.set_variable_length_onion_optional();
```

I'm a super fan of the explicative variable name, but I think that here the variable is too long? Mabey `onion_optional_features` is better?

vincenzopalazzo commented on lightning/src/routing/router.rs:
```diff
@@ -1330,7 +1332,7 @@ where L::Target: Logger {
 								ordered_hops.last_mut().unwrap().1 = node_info.features.clone();
 							} else {
-								ordered_hops.last_mut().unwrap().1 = NodeFeatures::empty();
+								ordered_hops.last_mut().unwrap().1 = variable_length_onion_optional_features.clone();
```
```suggestion
								ordered_hops.last_mut().unwrap().1 = onion_optional_features.clone();
```

TheBlueMatt reviewed (commented):
Basically lgtm, need to review the tests still, can you squash the fixup commits?

ViktorTigerstrom commented:
Fixed some documentation and updated the commit message. Sorry about that.

jkczyz commented on lightning/src/routing/router.rs:
```diff
@@ -2958,7 +2964,7 @@ mod tests {
 		assert_eq!(route.paths[0][3].fee_msat, 100);
 		assert_eq!(route.paths[0][3].cltv_expiry_delta, 42);
-		assert_eq!(route.paths[0][3].node_features.le_flags(), &Vec::<u8>::new()); // We dont pass flags in from invoices yet
+		assert_eq!(route.paths[0][3].node_features.le_flags(), &vec![0, 2]); // We dont pass flags in from invoices yet
```
Use `default_onion_format_flags`?

jkczyz commented on lightning/src/routing/router.rs:
```diff
@@ -1101,7 +1101,9 @@ where L::Target: Logger {
 	}
 
-	let empty_node_features = NodeFeatures::empty();
+	let mut onion_format_optional_features = NodeFeatures::empty();
+	onion_format_optional_features.set_variable_length_onion_optional();
```
I'd say use a method for this so it can be reused in the tests and name the variable `default_node_features`. No need to have the name talk about the feature itself. FWIW, I think we lose some meaning by dropping the "variable length" part of it, but we can avoid that by just using a generic name.

jkczyz commented on lightning/src/routing/router.rs:
```diff
@@ -1845,6 +1847,10 @@ mod tests {
 		}
 	}
 
+	fn default_onion_format_flags() -> Vec<u8> {
+		vec![0, 2]
+	}
```
Could drop this in favor of using the previously suggested method and calling `le_flags` on the result.

jkczyz reviewed (commented):
Largely look good but still need to review the tests while you address the comments.

ViktorTigerstrom commented:
Thanks for the feedback @jkczyz! Addressed it in the latest commit.

I've also realized that I should maybe expand my test to include routing through an unannounced node that isn't the target node, to cover that case as well. I'll add that if you think that it's of value to cover that case in the tests.

ViktorTigerstrom commented on lightning/src/routing/router.rs:
```diff
@@ -602,6 +602,17 @@ fn compute_fees(amount_msat: u64, channel_fees: RoutingFees) -> Option<u64> {
+///
+/// Default features are:
+/// * variable_length_onion_optional
+fn default_node_features() -> NodeFeatures {
```
Decided to also name this function to `default_node_features`, so that it possibly can be extended with more features in the future. I also considered if it would make more sense to add this as part of the `NodeFeatures` type itself, but figured that it probably belongs to the router context.

vincenzopalazzo replied:
I like this refactoring :)

ViktorTigerstrom replied:
Thanks!

TheBlueMatt commented on lightning/src/ln/onion_route_tests.rs:
```diff
@@ -577,6 +578,185 @@ fn test_onion_failure() {
+	// their `features` aren't used when creating the `route`.
+	network_graph.clear_nodes_announcement_info();
+
+	let scorer = test_utils::TestScorer::with_penalty(0);
+	let seed = [0u8; 32];
+	let keys_manager = test_utils::TestKeysInterface::new(&seed, Network::Testnet);
+	let random_seed_bytes = keys_manager.get_secure_random_bytes();
+	let read_only_network_graph = &network_graph.read_only();
+	let announced_route = get_route(
+		&origin_node.node.get_our_node_id(), &payment_params, read_only_network_graph,
+		Some(&origin_node.node.list_usable_channels().iter().collect::<Vec<_>>()),
+		1000000, TEST_FINAL_CLTV, origin_node.logger, &scorer, &random_seed_bytes).unwrap();
```
```suggestion
	let (announced_route, _, _, _) = get_route_and_payment_hash!(origin_node, nodes[3], 1_000_000);
```

TheBlueMatt commented on lightning/src/ln/onion_route_tests.rs:
```diff
@@ -577,6 +578,185 @@ fn test_onion_failure() {
+		htlc_maximum_msat: None,
+	}]);
+
+	let unannounced_chan_params = PaymentParameters::from_node_id(nodes[4].node.get_our_node_id()).with_route_hints(vec![last_hop]);
+	let unannounced_route = get_route(
+		&origin_node.node.get_our_node_id(), &unannounced_chan_params, read_only_network_graph,
+		Some(&origin_node.node.list_usable_channels().iter().collect::<Vec<_>>()),
+		10000, TEST_FINAL_CLTV, origin_node.logger, &scorer, &random_seed_bytes).unwrap();
```
```suggestion
	let (unannounced_route, _, _, _) = get_route_and_payment_hash!(nodes[0], nodes[4], last_hop, 10_000, TEST_FINAL_CLTV);
```

TheBlueMatt reviewed (commented):
Would be nice to use some of our macros to simplify the tests (I commented with two specific ones, but similar changes apply in `test_do_not_use_default_to_tlv_onions_if_other_features_not_supporting_them_exists` as well. Otherwise LGTM.

ViktorTigerstrom commented:
> Would be nice to use some of our macros to simplify the tests

Hmmm the current version of the `get_route_and_payment_hash` macro specifies that the `PaymentParameters` should be constructed using `InvoiceFeatures::known()`, which will then be used when creating the route for the last hop.
So I had to modify the `get_route_and_payment_hash` macro in order to be able to use it. However I'm not really sure if my modifications simplified things or just complicated the things more since the `macro` became more complicated to understand.

I therefore pushed the changes in a fixup commit, which I'll squash or drop depending on if you think the modifications are ok.

jkczyz commented on lightning/src/ln/functional_test_utils.rs:
```diff
@@ -1174,16 +1174,19 @@ macro_rules! get_route_and_payment_hash {
 			.with_route_hints($last_hops);
+		$crate::get_route_and_payment_hash!($send_node, $recv_node, payment_params, $recv_value, $cltv, true)
+	}};
+	($send_node: expr, $recv_node: expr, $payment_params: expr, $recv_value: expr, $cltv: expr, true /* with specified payment_params */) => {{
```
Would it be possible to just have this pattern (without the last `true`) and have uses of the previous pattern pass in the full `PaymentParameters`? I *think* there's only a few places in `priv_short_conf_tests.rs`.

ViktorTigerstrom replied:
Ok sure, I'll fix that ASAP!

ViktorTigerstrom commented:
Addressed the latest feedback and squashed the commits!

Note that I pushed the change of the `get_route_and_payment_hash` macro in a separate commit now as it wasn't really really related to the defaulting feature I've implemented, and I therefore wanted to avoid mixing the macro change with the feature's tests. Though I can of course squash them if you think that it's more appropriate!

jkczyz commented on lightning/src/ln/onion_route_tests.rs:
```diff
@@ -577,6 +577,170 @@ fn test_onion_failure() {
+
+#[test]
+fn test_do_not_use_default_to_tlv_onions_if_other_features_not_supporting_them_exists() {
+	// Tests that we do not default to creating tlv onions if any of these features exists when
+	// creating the specific hop in the route:
+	// 1. `InitFeatures` to the counterparty node exchanged with the init message to the node, which
+	// doesn't support variable length onions.
+	// 2. `NodeFeatures` in the `NodeAnnouncementInfo` of a node in sender node's the
+	// `network_graph`, which doesn't support variable length onions.
+	// 3. `InvoiceFeatures` specified by the receiving node, which doesn't support variable length
+	// onions, when no `NodeAnnouncementInfo` `features` exists for the receiver in the sender's
+	// `network_graph`.
```
The language could be simplified here by moving the "doesn't support variable length onions" part out of each item and said once preceding the list.

jkczyz commented on lightning/src/ln/onion_route_tests.rs:
```diff
@@ -577,6 +577,170 @@ fn test_onion_failure() {
+	// creating the specific hop in the route:
+	// 1. `InitFeatures` to the counterparty node exchanged with the init message to the node, which
+	// doesn't support variable length onions.
+	// 2. `NodeFeatures` in the `NodeAnnouncementInfo` of a node in sender node's the
```
drop trailing "the"

jkczyz commented on lightning/src/ln/onion_route_tests.rs:
```diff
@@ -577,6 +577,170 @@ fn test_onion_failure() {
+}
+
+#[test]
+fn test_do_not_use_default_to_tlv_onions_if_other_features_not_supporting_them_exists() {
```
To name somewhat similar to the previous test, how about: `test_do_not_default_to_onion_payload_tlv_format_when_unsupported`

jkczyz reviewed (commented):
Looks good. Just some nits on the tests.

vincenzopalazzo reviewed (approved):
LGTM

ViktorTigerstrom commented:
Thanks for the feedback @jkczyz! Addressed and squashed it with the last commit :)
